src/pathwayvis/components/reactionpanel/objectives/objectives.component.ts

```diff
@@ -0,0 +1,80 @@
+import * as angular from "angular";
+
+import { MapOptionService } from "../../../services/mapoption.service";
+import { ActionsService } from "../../../services/actions/actions.service";
+
+const DecafBiggProxy = 'https://api-staging.dd-decaf.eu/bigg/';
+
+class ObjectivesController {
+  public modelReactions = [];
+  public objectiveReaction: string[] = [];
+  private action: string = '';
+  private _mapOptions: MapOptionService;
+  private _actions: ActionsService;
+
+  constructor(mapOptions: MapOptionService,
+    actions: ActionsService) {
+    this._mapOptions = mapOptions;
+    this._actions = actions;
+
+    mapOptions.reactionsObservable.subscribe((reactions) => {
+      this.modelReactions = reactions;
+    });
+
+    mapOptions.objectiveReactionObservable.subscribe((reaction) => {
+      this.objectiveReaction = [];
+      if (reaction !== '') this.objectiveReaction.push(reaction);
+    });
+  }
+
+  public queryModelReactions(query: string): any[] {
+    query = query.toLowerCase();
+    return this.modelReactions.filter((reaction) => {
+      return reaction.name.toLowerCase().includes(query)
+        || reaction.id.toLowerCase().includes(query);
+    });
+  }
+
+  public setAsObjectiveSelectedItem(item) {
+    if (!item) return;
+    const doObjectiveAction = this._actions.getAction('reaction:objective:do');
+    this._mapOptions.setObjectiveReaction(item.id);
+    this._mapOptions.actionHandler(doObjectiveAction, { id: item.id })
+    .then(this.updateMapData.bind(this));
+  }
+
+  public onUndoClick(selectedReaction: string): void {
+    const undoObjectiveAction = this._actions.getAction('reaction:objective:undo');
+    this._mapOptions.setObjectiveReaction('');
+    this.objectiveReaction = [];
+    this._mapOptions.actionHandler(undoObjectiveAction, { id: selectedReaction })
+    .then(this.updateMapData.bind(this));
+  }
+
+  public objectiveDisplay(item) {
+    return item;
+  }
+
+  private updateMapData(response) {
+    this._mapOptions.setCurrentGrowthRate(parseFloat(response['growth-rate']));
```

Comment:
I see this bit of code repeating everywhere in the code. This might be challenging, but could you check if it's feasible the do this only in one place?